### PR TITLE
Migrate Device Controller from transport layer to exchange layer and enable exchange layer by default.

### LINKED ISSUE #5625: Migrate Device Controller from transport layer to messaging layer and enable ExchangeManager by default.
#### Problem
Currently, Device Controller is still using SecureSessionMgr to send and receive ZCL messages, and ExchangeManager is only enabled in integration test apps such as Echo Requester and Echo Responder. Therefore, all protocols (IM, Echo, BDX ...) are disabled in production apps.

This prevent us from testing all features based on CHIP protocol, such as cluster based networking provisioning.

 #### Proposed Solution
1. Using exchange in Device Controller to send and receive messages and decouple it from SecureSessionMgr. 
2. Introduce a new protocol DeviceManagement to send/receive legacy ZAL messages over DeviceController

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
Currently, Device Controller is still using SecureSessionMgr to send and receive ZCL messages, and ExchangeManager is only enabled in integration test apps such as Echo Requester and Echo Responder. Therefore, all protocols (IM, Echo, BDX ...) are disabled in production apps.

This prevent us from testing all features based on CHIP protocol, such as cluster based networking provisioning.

 #### Summary of Changes
1. Using exchange in Device Controller to send and receive messages and decouple Device Controller from SecureSessionMgr.
2. Introduce a new protocol DeviceManagement to send/receive legacy ZAL messages over DeviceController

 Fixes #5625, #5493

 #### Test
 On Server Side:
 ./chip-all-clusters-app

On Client Side:
./chip-tool pairing bypass 192.168.86.157 11097
./chip-tool onoff on 1

Confirm all cluster messages can be sent and received.

### CODE REVIEW

**Comment by bzbarsky-apple** [comment]
@yufengwangca Are the codesize results expected?  What's the 4KB table we are adding in bss?  And the 1.5KB rodata?

Is the 8 KB of .text expected?  That's a pretty substantial increase...

File: src/app/util/chip-message-send.cpp
Comment (bzbarsky-apple) [inline src/app/util/chip-message-send.cpp:90]: Why do we need to set this initiator flag, as an API consumer?  Shouldn't the ExchangeContext know whether we're the initiator or not?  It sure looks like it does, and it ignores the "kFromInitiator" flag, as it should.  In fact, absolutely nothing uses that flag.   I filed https://github.com/project-chip/connectedhomeip/issues/5651 to do that.

Why do we want to force no ack here?  I don't see a reason for it....

File: src/controller/CHIPDevice.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:59]: Why did this get moved?  There's no obvious reason to move it, and it makes review much more difficult (have to check it line by line for differences, etc).  Please put it back where it used to be.

File: src/controller/CHIPDevice.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:118]: That seems like it would be pretty common.  Can you please file a followup to have the flags arg be optional, with that value as default?

File: src/controller/CHIPDevice.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:133]:
```suggestion
        sendFlags.Set(Messaging::SendMessageFlags::kFromInitiator).Set(Messaging::SendMessageFlags::kNoAutoRequestAck);
```

Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:133]: Why are we setting these flags on _resend_ but not on initial send?  As above in chip-send, I don't think we should set these at all.

File: src/controller/CHIPDevice.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:87]: Why did this get moved too?

File: src/controller/CHIPDevice.h
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.h:94]:
```suggestion
     * @param[in] msgType     The message type of the message to be sent.  Must be a valid message type for protocolId.
```

File: src/controller/CHIPDevice.h
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.h:95]:
```suggestion
     * @param[in] message     The message payload to be sent.
```

File: src/controller/CHIPDevice.h
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.h:99]: We should add a template helper that infers the protocol id from a strongly typed message type....  and cosider making the untyped method protected.  Followup is fine for this.

File: src/controller/CHIPDevice.h
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.h:186]: I don't think that makes sense.  Why can we get away with not having an exchange manager here?

**Comment by yufengwangca** [comment]
> @yufengwangca Are the codesize results expected? What's the 4KB table we are adding in bss? And the 1.5KB rodata?
> 
> Is the 8 KB of .text expected? That's a pretty substantial increase...

Yes, since this is our first time to enable messaging layer (ExchangeContext/ExchangeMgr/Protocols) on real examples. Right now, messaging layer is only enabled on several test apps (Echo/IM) for integration test.

File: src/controller/CHIPDeviceController.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:138]: Don't we need to free this in Shutdown()?  I don't see us doing it there.

File: src/lwip/qpg6100/lwipopts.h
Comment (bzbarsky-apple) [inline src/lwip/qpg6100/lwipopts.h:92]: This does not seem obviously related to the rest of these changes...

File: src/messaging/ExchangeMgrDelegate.h
Comment (bzbarsky-apple) [inline src/messaging/ExchangeMgrDelegate.h:50]: "new" connection?

File: src/transport/tests/TestSecureSessionMgr.cpp
Comment (bzbarsky-apple) [inline src/transport/tests/TestSecureSessionMgr.cpp:163]: We should set a useful protocol/message-id on this payload header.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

**Comment by bzbarsky-apple** [comment]
> Yes, since this is our first time to enable messaging layer (ExchangeContext/ExchangeMgr/Protocols) on real examples.

I understand that _some_ increase is expected.  I am asking whether this specific (fairly large!) magnitude is expected.

Put another way: this PR is using something like 4% of the last flash target I have seen for CHIP, and 6% of the last RAM target I saw.  This seems like a lot.  Why is it using that much?

**Comment by bzbarsky-apple** [comment]
OK, I did some poking around at the specific symbols involved:

* `gExchangeMgr` is 4KB of bss.  Presumably for whatever tables it's got inside it, but that seems like a fair amount of storage...  It's worth looking into what's in there and whether it needs to be this big.
* The 2KB of rodata seems to be generally strings in various functions.  For example, 200 bytes of strings in `ExchangeManager::OnMessageReceived`.  Presumably from the various logging.  Maybe that will go away automatically in production?  Would be good to measure codesize in that configuration.  And if not, we should think about having shorter strings.
* We're pulling in a bunch of new code for those 8KB of codesize.  Not just ExchangeContext and ExchangeManager, but also ReliableMessageManager, message counter bits, etc.  This is the drawback of not turning things on incrementally..... `ExchangeContext` on its own is 2.5KB; `ExchangeManager` is 2KB.  `MessageCounterSyncMgr` is 1.8KB, `ReliableMessageMgr` is 1.4KB.  That's most of the codesize increase..... 
 
 
For comparison, our entire mdns impl is something like 11KB, and even that might be more than we want.  So it might be worth looking into why these are so big.

File: src/protocols/Protocols.h
Comment (yufengwangca) [reply src/protocols/Protocols.h]: Acked.

File: src/app/server/SessionManager.h
Comment (yufengwangca) [reply src/app/server/SessionManager.h]: None of them should exist in this header, their implementation parts are in Server.c, moved the declarations Server.h

File: src/controller/CHIPDevice.cpp
Comment (yufengwangca) [reply src/controller/CHIPDevice.cpp]: There are two SendMessage, this PR delete the first one, which cause the confusion on review, merged the second one to the first one solve this problem.

File: src/controller/CHIPDevice.h
Comment (yufengwangca) [reply src/controller/CHIPDevice.h]: This Init is used by RendezvousSession, which still run on SecureSessionMgr directly. ExchangeMgr is not used by SecureSessionMgr. We will replace legacy pairing with clustered based network provisioning, this Init function will be deleted together with RendezvousSession. Right now, just init ExchangeMgr here to keep consistency.

File: src/lwip/qpg6100/lwipopts.h
Comment (yufengwangca) [reply src/lwip/qpg6100/lwipopts.h]: This is to pass CI since we reached the memory limit on qpg6100, this change has already been merged with https://github.com/project-chip/connectedhomeip/pull/5610.

File: src/transport/tests/TestSecureSessionMgr.cpp
Comment (yufengwangca) [reply src/transport/tests/TestSecureSessionMgr.cpp]: Acked

File: src/messaging/ExchangeMgrDelegate.h
Comment (yufengwangca) [reply src/messaging/ExchangeMgrDelegate.h]: Acked

File: src/controller/CHIPDevice.cpp
Comment (yufengwangca) [reply src/controller/CHIPDevice.cpp]: Same reason as above

File: src/controller/CHIPDevice.cpp
Comment (yufengwangca) [reply src/controller/CHIPDevice.cpp:82]: https://github.com/project-chip/connectedhomeip/issues/5670

File: src/lwip/qpg6100/lwipopts.h
Comment (bzbarsky-apple) [reply src/lwip/qpg6100/lwipopts.h]: Great, thank you for spinning that off into a separate PR!

File: src/controller/CHIPDevice.h
Comment (yufengwangca) [reply src/controller/CHIPDevice.h:99]: https://github.com/project-chip/connectedhomeip/issues/5672

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp:138]: Acked.

File: src/app/util/chip-message-send.cpp
Comment (yufengwangca) [reply src/app/util/chip-message-send.cpp]: Here we defined a temporary Protocol TempZCL to carry over legacy ZCL messages, but unlike other standard protocols like Interaction Model, Echo ..., there is no protocol instances to host the conversation with exchange, such as EchoClient and EchoServer. 

Therefore, we used ephemeral exchange to send and receive ZCL messages and TempZCL Protocol does not know the semantics of each message. This is similar to the current implementation, just use an ephemeral exchange to send message and use its unsolicited message handler to receive messages.

We need to set flag kFromInitiator to allow receiver to deliver this message to corresponding unsolicited message handlers, and we also need to set flag kNoAutoRequestAck since there is no dedicated conversation to host the ack message.

File: src/controller/CHIPDevice.cpp
Comment (yufengwangca) [reply src/controller/CHIPDevice.cpp]: All messages carried over TempZCL protocol should have kFromInitiator and kNoAutoRequestAck flags set as I explained above.

File: src/app/util/chip-message-send.cpp
Comment (bzbarsky-apple) [reply src/app/util/chip-message-send.cpp]: Hmm...  I can see this being ok very short-term, but it seems to me we should actually have a protocol instance so we can actually send the response back on the right exchange.  I'm OK with that happening in a followup, but:

1. There should be clear comments in the code explaining why we are doing these weird things, more or less summarizing the explanation here.
2. Those comments should reference an issue that we resolve in the near future which will remove he need for doing those weird things.

File: src/controller/CHIPDevice.cpp
Comment (bzbarsky-apple) [reply src/controller/CHIPDevice.cpp]: This explanation should be in the code.

File: src/app/util/chip-message-send.cpp
Comment (yufengwangca) [reply src/app/util/chip-message-send.cpp]: Added. We already have the protocol instance implemented in IM protocol for this purpose. After we convert all legacy ZCL Ember messages to IM constructs, this logic will be cleaned up.

**Comment by yufengwangca** [comment]
> OK, I did some poking around at the specific symbols involved:
> 
> * `gExchangeMgr` is 4KB of bss.  Presumably for whatever tables it's got inside it, but that seems like a fair amount of storage...  It's worth looking into what's in there and whether it needs to be this big.
> * The 2KB of rodata seems to be generally strings in various functions.  For example, 200 bytes of strings in `ExchangeManager::OnMessageReceived`.  Presumably from the various logging.  Maybe that will go away automatically in production?  Would be good to measure codesize in that configuration.  And if not, we should think about having shorter strings.
> * We're pulling in a bunch of new code for those 8KB of codesize.  Not just ExchangeContext and ExchangeManager, but also ReliableMessageManager, message counter bits, etc.  This is the drawback of not turning things on incrementally..... `ExchangeContext` on its own is 2.5KB; `ExchangeManager` is 2KB.  `MessageCounterSyncMgr` is 1.8KB, `ReliableMessageMgr` is 1.4KB.  That's most of the codesize increase.....
> 
> For comparison, our entire mdns impl is something like 11KB, and even that might be more than we want. So it might be worth looking into why these are so big.

Thanks for the memory analysis, Yes, "enable messaging layer" means we pull in all functions within the messaging layer, not just ExchangeContext and ExchangeMgr classes. 

I used to have a PR https://github.com/project-chip/connectedhomeip/pull/4313 to enable ExchangeMgr by default to avoid last minute surprise. But it is still not too late now, this PR will force us pay attention to the memory budget increase as we land new features in messaging layer.

File: src/app/server/Server.cpp
Comment (kghost) [inline src/app/server/Server.cpp:321]: I strongly suggest removing `packetHeader` and `payloadHeader` arguments.

`packetHeader` is used to retrieve node id, but we can get node id from `ExchangeContext` which is more reliable.

`payloadHeader` is only used to get protocol id of the packet, I think we can use protocol id and to replace payloadHeader.

Comment (yufengwangca) [reply src/app/server/Server.cpp:321]: This function prototype is to implement the callbacks of ExchangeDelegate interface,  changing this interface will impact all ExchangeContext clients. 

If I understand correctly, you are talking about refactor current ExchangeDelegate interface in general, right? I suggest we register a separate ticket to discuss this

**Comment by yufengwangca** [comment]
The following are major components which contribute the size increase:

Flash rodata:  

ReliableMessageContext/ReliableMessageMgr
Total: +670

MessageCounterSyncMgr
Total: +547

ExchangeManager
Total: +523

ExchangeContext
Total: +130

------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Flash code

ReliableMessageContext/ReliableMessageMgr
Total: +2250

ExchangeManager
Total: +1783

MessageCounterSyncMgr
Total: +1656

ExchangeContext
Total: +1464

**Comment by yufengwangca** [comment]
1) Optimize the RAM usage by auditing configuration constants of messaging layer. Now the RAM increase of this PR has changed from +4096 to +2907

2) Filed an issue to track reducing the .rodata usage from logging
strings, especially in production configurations.  This can be a general
thing we look at, not just for this PR.

https://github.com/project-chip/connectedhomeip/issues/5740

3) Filed an issue to track reducing the .text usage of the code. 
https://github.com/project-chip/connectedhomeip/issues/5741

**Review by pan-apple — APPROVED** [review]
Code changes in the PR look good. Some follow up work is needed to optimize flash usage. Maybe in a separate PR.

**Review by andy31415 — APPROVED** [review]

**Review by yunhanw-google — APPROVED** [review]

**Review by Damian-Nordic — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp
Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp:223]: Yes, apparently not needed.  But also, why here is `mSessionMgr->SetDelegate(nullptr);` called here, but above it wasn't?

Is that missing from above?  Is this a bug?

**Review by msandstedt — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp
Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp:220]: OK, this seems to be resolved now in 923b110d46ea55a5655b487c3152ddb33db2938e.

The delete of `mTransportMgr` has now also been removed.  Is that correct?

Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp:220]: I guess this is caused by yesterday's rebase, we didn't release those managers before, it looks another PR recently landed also cover this problem.

mSessionMgr->SetDelegate(nullptr); is not needed since we deleted mSessionMgr as a whole. I would leave  mStorageDelegate->SetStorageDelegate(nullptr); there since mStorageDelegate is not allocated/released within this class.

**Comment by yunhanw-google** [comment]
@bzbarsky-apple @andy31415 we need this PR to unblock enablement for interaction mode command integration in example code, please help merge it if possible, thanks

File: src/controller/CHIPDevice.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDevice.cpp:109]: If we reach the resend codepath and it hits a failure for any of its `ReturnErrorOnFailure` calls, we will skip doing this step and end up leaking an exchange, right?

I'm OK with a followup to track this, just so we don't keep blocking this PR, but please file an issue to address this (e.g. by adding some sort of RAII exchange-closing class).

File: src/messaging/ReliableMessageProtocolConfig.h
Comment (bzbarsky-apple) [inline src/messaging/ReliableMessageProtocolConfig.h:96]: I wonder whether as a followup we should use the min of pool size and `MAX_EXCHANGE_CONTEXTS` for these.....

File: src/protocols/Protocols.h
Comment (bzbarsky-apple) [inline src/protocols/Protocols.h:82]:
```suggestion
CHIP_STANDARD_PROTOCOL(TempZCL, 0x0009)             // For carrying ZCL-compatible messages, to be removed when we no longer use them
```

File: src/controller/CHIPDevice.cpp
Comment (bzbarsky-apple) [reply src/controller/CHIPDevice.cpp:109]: Filed https://github.com/project-chip/connectedhomeip/issues/5922

File: src/messaging/ReliableMessageProtocolConfig.h
Comment (bzbarsky-apple) [reply src/messaging/ReliableMessageProtocolConfig.h:96]: Filed https://github.com/project-chip/connectedhomeip/issues/5924

**Review by bzbarsky-apple — APPROVED** [review]

File: src/protocols/Protocols.h
Comment (bzbarsky-apple) [inline src/protocols/Protocols.h:82]:
```suggestion
 CHIP_STANDARD_PROTOCOL(TempZCL, 0x0009) // For carrying ZCL-compatible messages, to be removed when we no longer use them
```

File: src/protocols/Protocols.h
Comment (bzbarsky-apple) [inline src/protocols/Protocols.h:82]:
```suggestion
CHIP_STANDARD_PROTOCOL(TempZCL, 0x0009) // For carrying ZCL-compatible messages, to be removed when we no longer use them
```

**Review by mspang — APPROVED** [review]

**Pull request merged**
